src/analysis/cfg.h

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2023 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef wasm_analysis_cfg_h
+#define wasm_analysis_cfg_h
+
+#include <iostream>
+#include <vector>
+
+#include "wasm.h"
+
+namespace wasm::analysis {
+
+struct BasicBlock;
+
+struct CFG {
+  // Iterate through basic blocks.
+  struct iterator;
+  iterator begin() const;
+  iterator end() const;
+  size_t size() const;
+
+  static CFG fromFunction(Function* func);
+
+  void print(std::ostream& os, Module* wasm = nullptr);
+
+private:
+  std::vector<std::vector<Expression*>> blocks;
+  std::vector<std::vector<size_t>> succs;
+  std::vector<std::vector<size_t>> preds;
```

Comment:
> I'm guessing that these last two vectors are over basic block indexes? Perhaps declaring a type for that index would make it clearer, `using BlockIndex = Index` ?

Yes. I'll switch to using `Index` and add comments. I don't think a `BlockIndex` type alias would add much over that.

> (`Index` instead of `size_t` makes sense to me because the number of blocks is limited by the number of items in a wasm module, which `Index` represents; and it is smaller and more efficient than `size_t`).

👍 

> Separately, why indexes and not direct pointers?

I turned to indices instead of pointers because it takes more effort (and sometimes allocations) to maintain stable addresses. OTOH, this CFG utility is meant to be immutable after it's constructed, so using pointers wouldn't require separate allocations if we were careful. Even if we did use pointers, though, we would still want to be able to recover the indices for printing, at least.

Reply:
This code no longer exists.